pkg/devicetwin/dtcontroller_test.go

```diff
@@ -0,0 +1,757 @@
+/*
+Copyright 2018 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package devicetwin
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"reflect"
+	"testing"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/golang/mock/gomock"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtclient"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtmodule"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dttype"
+)
+
+//TestInitDTController is function to test InitDTController().
+func TestInitDTController(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name      string
+		context   *context.Context
+		want      *DTController
+		wantError error
+	}{
+		{
+			name:    "InitDTControllerActualContextTest",
+			context: mainContext,
+			want: &DTController{
+				HeartBeatToModule: make(map[string]chan interface{}),
+				DTContexts:        dtContexts,
+				DTModules:         make(map[string]dtmodule.DTModule),
+				Stop:              make(chan bool, 1),
+			},
+			wantError: nil,
+		},
+		{
+			name:    "InitDTControllerEmptyContextTest",
```

Comment:
what is the difference between this test and previous one ? do we have a failure scenario ? if yes , let's introduce that instead.

Reply:
One will just return an empty context....The function always returns the error field as nil hence it is not possible to have a failure scenario.